### DISCUSSION TITLE
login route with users data

### DISCUSSION THREAD

**Discussion by mobergmann** [maxcountryman/axum-login#93] [Q&A] [answered]
I want to log in the user, whichever calles the login route. I searched the examples, but I only found the usage of `get_rusty_user` and `let user: User = sqlx::query_as("select * from users where id = 1")` to retrieve the user's login data. I am new to axum (and rust), so I am may overseeing something. How can I retrieve the User data in the login route, e.g. like in my `sign_up` route with the usage of a client sending a request (e.g. in JavaScript using the `XMLHttpRequest`) (or a html-form, which isn't my use case)?
```rust
async fn sign_up(Json(payload): Json<User>) -> impl IntoResponse { ... }
```

My sign-in route so far:
```rust
async fn sign_in(mut auth: AuthContext) {
    let u = User { ... };
    auth.login(&u).await.unwrap();
}
```
what I imagine retrieving a user in the login route would look like:
```rust
async fn sign_in(Json(payload): Json<User>, mut auth: AuthContext) {
    auth.login(&payload).await.unwrap();
}
```

**Comment by maxcountryman** [comment, marked as answer]
The assumption here is your sign in route would take in some kind of unique identifier and some way of verifying it's being used in an authorized way.

So for example, you might identify users by their email address. Your route would receive their email address and then find the user in your store.

However, it's important to point out that this isn't enough to know whoever sending along the email address is entitled to that user's identity and so you will also likely want additional context, such as a password.

In your application you then need to do two things:

1. Find the user in your store
2. And verify the provided credentials match your expectations.

Your route could look something like this (this is incomplete and won't compile, but hopefully you get some idea):

```rust
#[derive(Debug, Default, Clone, sqlx::FromRow)]
struct User {
    email: String,
    password_hash: String,
}

impl AuthUser<String> for User {
    fn get_id(&self) -> String {
        self.email
    }

    fn get_password_hash(&self) -> SecretVec<u8> {
        SecretVec::new(self.password_hash.clone().into())
    }
}

type AuthContext = axum_login::extractors::AuthContext<String, User, SqliteStore<User>>;

#[derive(Deserialize)]
struct UserSignIn {
    email: String,
    password: String,
}

async fn sign_in(mut auth: AuthContext, Json(UserSignIn { email, password }): Json<UserSignIn>) {
        let pool = SqlitePoolOptions::new()
            .connect("sqlite/user_store.db")
            .await
            .unwrap();
        let user = sqlx::query_as!(User, "select * from users where email = ?", email)
            .fetch_one(&pool)
            .await
            .unwrap();
        // Check password is correct by comparing auth.get_password_hash() with the provided password; use e.g. bcrypt, etc for this.
        auth.login(&user).await.unwrap(); // only call this AFTER the password has been checked.
}
```

**Reply by mobergmann** [reply]
when using an email (String) as the id, wouldn't then the `AuthContext` be `axum_login::extractors::AuthContext<String, User, SqliteStore<User>>;`?

**Reply by mobergmann** [reply]
Is the order of the `auth` parameter important, as [this](https://github.com/maxcountryman/axum-login/discussions/28#discussioncomment-6486881) post states?

**Reply by maxcountryman** [reply]
Yes, String is the correct type for id in that case and AuthContext needs to reflect that.

**Reply by maxcountryman** [reply]
It's also correct that Json should go last because it consumes the body. That's related to Axum.
